### ISSUE TITLE
Implement DTSL (encrypted UDP sockets)

### ISSUE THREAD

**Issue by beyonlo** [micropython/micropython#5270]
Hello.

Are there any example about to use DTLS? I would like to use UDP protocol to communicate, but in the secure mode, like as a TCP socket using TLS. I found on docs just about ussl for TCP sockets.

Thank you.

**Comment by MrSurly** [comment]
I don't believe it's implemented.  A while back, I [took a look](https://github.com/micropython/micropython-esp32/issues/188) at doing so, and despite what I said in that thread, I did indeed abandon it.

IIRC, the underlying stack (mbedtls) supports this.

**Comment by beyonlo** [comment]
@dpgeorge Any plans to implement DTLS?

Thank you.

**Comment by dpgeorge** [comment]
>  Any plans to implement DTLS?

Not in the short term, but in the medium term (~3 months) it could be done if there's enough interest.

**Comment by dpgeorge** [comment]
For reference, using DTLS on CPython is possible via https://github.com/rbit/pydtls

**Comment by dmartauz** [comment]
I am also interested in DTLS support, however I would prefer it to be done similarly to ucryptolib.aes with encrypt/decrypt methods. The reason is that the output would be used as UDP payload for NB-IoT modem controlled via AT commands which is not bound to lwip.

**Comment by dpgeorge** [comment]
> The reason is that the output would be used as UDP payload for NB-IoT modem controlled via AT commands which is not bound to lwip.

Note that the `ssl` module can currently be used stand-alone, ie without lwip/sockets.  It just requires a suitable stream (eg user class derived from `io.IOBase`) to be passed in to `ssl.wrap_socket`  I would envisage a similar thing for DTLS.

**Comment by Sebbb** [comment]
> > Any plans to implement DTLS?
> 
> Not in the short term, but in the medium term (~3 months) it could be done if there's enough interest.

I'm interested as well, I'm searching for an option to encrypt/decrypt UDP packets on an ESP32.

The possibility to communicate securely definitely makes sense, also in the IoT/embedded world :)
